cwl_tes/ftp.py

```diff
@@ -149,8 +151,22 @@ def open(self, fn, mode):
             return super(FtpFsAccess, self).open(fn, mode)
         if 'r' in mode:
             host, user, passwd, path = self._parse_url(fn)
-            handle = urllib.request.urlopen(
-                "ftp://{}:{}@{}/{}".format(user, passwd, host, path))
+            url = "ftp://{}:{}@{}/{}".format(user, passwd, host, path)
+            # ftp = self._connect(fn)
+            with NamedTemporaryFile(mode='wb', delete=False) as dest:
+                c = pycurl.Curl()
+                c.setopt(c.URL, url)
+                c.setopt(c.WRITEDATA, dest)
+                c.perform()
+                response_code = c.getinfo(c.RESPONSE_CODE)
+                if not (200 <= response_code <= 299):
+                    print("There was a problem downloading " +
+                          f"{c.getinfo(c.EFFECTIVE_URL)} ({response_code})")
+                c.close()
+                temp_fname = dest.name
+
+            # Return a file handle in read mode
+            handle = open(temp_fname, mode)
```

Comment:
This will end up creating a lot of files in `/tmp` over the course of different cwl-tes runs. Judging from the fact that this method returns a handle, you can use a context manager that will delete the underlying file when the caller is done with it:
```python
from contextlib import contextmanager

def use_and_delete(fname, mode='r'):
    try:
        with open(fname, mode=mode) as fp:
            yield fp
    finally:
        os.unlink(fname)
```
and then in your code:
```
handle = use_and_delete(temp_fname)
return handle
```

Reply:
In the test run this produces:
```
ERROR [job Md5sum] job error:
("Error collecting output for parameter 'output_file':\nmd5.cwl:32:3: __enter__", {})
```

Reply:
I should have added that I didn't actually test the code :) This should work, at least in principle:
```python
from contextlib import contextmanager

@contextmanager  # <--- this was missing
def use_and_delete(fname, mode='r'):
    try:
        with open(fname, mode=mode) as fp:
            yield fp
    finally:
        os.unlink(fname)

```

This can then be used as
```python
handle = use_and_delete("myfile.txt")

with handle as fp:
   # use fp as a regular file handle. 
```

I am not sure that this is how the `handle` object is used in the rest of the code, though.

Reply:
Feel free to use another approach, though. The idea is just to make sure that temporary files get properly cleaned up.

Reply:
That works! I have added this to the PR branch.